### PR TITLE
Allow object managers to cancel unnecessary object pull requests.

### PR DESCRIPTION
**Pull request by robertnishihara**
This is very much a work in progress, but I'm opening it now to try to get some earlier feedback on the best design here.

Note that some related work was recently done in https://github.com/ray-project/ray/pull/3276, which did some deduplication on the "sender" side. That is, no object manager will send the same object multiple times to the same remote object manager within a certain window of time.

This PR should address the **N^2** broadcast issue in #2945. Ideally, when broadcasting an object to N other nodes, the object would be transmitted exactly N times. This PR doesn't quite achieve that, e.g., I still see it transmitted around **2N** times.

__In this PR:__
- When an object manager is trying to fetch an object, it may request it from multiple (potentially all) other remote object managers that have the object.
- As soon as the object manager begins receiving an object from a remote object manager, it will send messages to all other object managers that it has requested the object from (except for the one that is currently transmitting the object) and cancel those requests.
- Object managers maintain a queue of "pending object pushes". These are pushes that have not been posted to one of the "asio send services". If an object manager receives a cancellation message, it will look in the queue.
- When one of the send service threads finishes sending an object, we see if we can dispatch more sends.

__Some notes/questions:__
- **Chunks versus objects.** We queue "object transfers" in the `pending_pushes_` queue. But when we actually decide to push an object, we break it into chunks and post the sends of the individual chunks to the send service threads. So objects are queued in the object manager queue and chunks are queued in the asio queue. It may be advantageous to queue chunks in the object manager queue because that way we could cancel an object transfer even after we've sent some of its chunks. The eventual solution should ideally allow for an object manager to receive different chunks from different object managers, though that will probably require more thought.
- **Data structures.** Currently I'm using something very naive, for `pending_pushes_` (a deque). The operations that we need are to be able to pop off the front, append to the back, and remove from anywhere. So basically an "ordered map" or something like that.

@guoyuhong @stephanie-wang @elibol @ujvl @ericl

### CODE REVIEW

File: src/ray/object_manager/object_manager.h [hunk 1]
```diff
@@ -416,6 +445,14 @@ class ObjectManager : public ObjectManagerInterface {
   /// remote object managers.
   std::unordered_map<ObjectID, PullRequest> pull_requests_;
 
+  /// A set of the chunks that are in the process of being pushed to remote
+  /// object managers. Each item consists of the ID of the remote object
+  /// manager, the ID of the object, and the index of the chunk.
+  std::vector<std::tuple<ClientID, ObjectID, uint64_t>> in_progress_pushes_;
```
Comment (ericl) [inline src/ray/object_manager/object_manager.h]: Can this be a map? Also, add the size of this to the class debug string.

File: src/ray/object_manager/object_manager.h [hunk 2]
```diff
@@ -416,6 +445,14 @@ class ObjectManager : public ObjectManagerInterface {
...
+
+  /// The objects that are waiting to be pushed to other machines.
+  std::deque<std::pair<ObjectID, ClientID>> pending_pushes_;
+
```
Comment (ericl) [inline src/ray/object_manager/object_manager.h]: Add the size of this to debug string.

File: src/ray/object_manager/object_manager.cc [hunk 3]
```diff
@@ -800,29 +884,67 @@ void ObjectManager::ReceivePullRequest(std::shared_ptr<TcpClientConnection> &con
...
-        });
 
-  });
+  // Now that we are about to start receiving this object, if we had requested
+  // it from other object managers, cancel those requests.
+  auto it = pull_requests_.find(object_id);
+  if (it != pull_requests_.end()) {
+    for (auto const &requested_client_id : it->second.clients_requested) {
+      RAY_CHECK(requested_client_id != client_id_);
+      SendCancelPullRequest(object_id, requested_client_id);
```
Comment (ericl) [inline src/ray/object_manager/object_manager.cc]: Add the total count of cancels etc. to debug string?

File: src/ray/object_manager/object_manager.h (hunk 2 above)
Comment (robertnishihara) [reply src/ray/object_manager/object_manager.h]: Great suggestion, but fortunately I already did this :)

**Review by stephanie-wang — COMMENTED** [review]
Some high-level comments:
1. I would abstract `pending_pushes_` and `in_progress_pushes_` out into a separate class, something like `PushQueue`.
2. I would make `pending_pushes_` a `std::list` that keeps the FIFO ordering that the push requests were received in. Then, I would add a `std::unordered_map<std::pair<ClientID, ObjectID>, std::list::iterator>` to deduplicate pushes.
3. I wonder if we should make `PushQueue` (assuming we do make it a separate class) thread-safe instead of coordinating threads with `asio::io_service::post`? It'd be great to benchmark that ASAP.

File: src/ray/object_manager/object_manager.cc [hunk 4]
```diff
@@ -776,18 +873,40 @@ void ObjectManager::ConnectClient(std::shared_ptr<TcpClientConnection> &conn,
...
+
   connection_pool_.RemoveReceiver(conn);
 
   // We don't need to clean up unfulfilled_push_requests_ because the
   // unfulfilled push timers will fire and clean it up.
+
+  // Clean up the bookkeeping of the pushes that were in the process of being
+  // pushed to this client.
+  for (auto it = in_progress_pushes_.begin(); it != in_progress_pushes_.end();) {
+    if (std::get<1>(*it) == client_id) {
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.cc]: why not `it->second`?

**Comment by robertnishihara** [comment]
Ok, this PR unfortunately seems to introduce overhead for fetching 1000 small objects from a remote machine. I'll need to investigate that more.

File: src/ray/object_manager/object_manager.cc (hunk 4 above)
Comment (robertnishihara) [reply src/ray/object_manager/object_manager.cc]: It's a tuple, not a pair.

**Pull request closed without merging**